File: Darts/src/code/ai/DartsSimulationGolf.java
package code.ai;

import code.db.GameEntity;
import code.object.Dart;
import code.screen.Dartboard;
import util.Debug;

public final class DartsSimulationGolf extends AbstractDartsSimulation
{
	private static final int ROUNDS = 18;
	
	private int score = 0;
	
	public DartsSimulationGolf(Dartboard dartboard, AbstractDartsModel model)
	{
		super(dartboard, model);
	}

	@Override
	public boolean shouldPlayCurrentRound()
	{
		return currentRound <= ROUNDS;
	}
	
	@Override
	protected void resetVariables()
	{
		super.resetVariables();
		score = 0;
	}
	
	@Override
	public void startRound()
	{
		resetRound();
		
		int dartNo = dartsThrown.size() + 1;
		model.throwGolfDart(currentRound, dartNo, dartboard);
	}
	
	private void finishedRound()
	{
		hmRoundNumberToDarts.put(currentRound, dartsThrown);
		
		Dart drt = dartsThrown.lastElement();
		int roundScore = drt.getGolfScore(currentRound);
		score += roundScore;
		
		Debug.appendBanner("Round " + currentRound, logging);
		Debug.append("Darts [" + dartsThrown + "]", logging);
		Debug.append("Score [" + roundScore + "]", logging);
		Debug.append("Total Score [" + score + "]", logging);
		
		currentRound++;
	}
	
	@Override
	public void dartThrown(Dart dart)
	{
		dartsThrown.add(dart);
		
		int noDarts = dartsThrown.size();
		int stopThreshold = model.getStopThresholdForDartNo(noDarts);
		
		if (noDarts == 3
		 || dart.getGolfScore(currentRound) <= stopThreshold)
		{
			finishedRound();
		}
		else
		{
			int dartNo = dartsThrown.size() + 1;
			model.throwGolfDart(currentRound, dartNo, dartboard);
		}
	}

	@Override
	public int getTotalScore()
	{
		return score;
	}

	@Override
	public String getGameParams()
	{
		return "" + ROUNDS;
	}
	
	@Override
	public int getGameType()
	{
		return GameEntity.GAME_TYPE_GOLF;
	}

}

File: src/main/kotlin/dartzee/ai/DartsSimulationGolf.kt
package dartzee.ai

import dartzee.db.PlayerEntity
import dartzee.game.GameType
import dartzee.`object`.Dart

private const val ROUNDS = 18

class DartsSimulationGolf(player: PlayerEntity, model: DartsAiModel) : AbstractDartsSimulation(player, model)
{
    override val gameType = GameType.GOLF
    override val gameParams = "$ROUNDS"
    var score = 0

    override fun shouldPlayCurrentRound(): Boolean
    {
        return currentRound <= ROUNDS
    }

    override fun resetVariables()
    {
        super.resetVariables()
        score = 0
    }

    override fun getTotalScore() = score

    override fun startRound()
    {
        resetRound()

        val dartNo = dartsThrown.size + 1
        val pt = model.throwGolfDart(currentRound, dartNo)
        dartThrown(pt)
    }

    private fun finishedRound()
    {
        confirmRound()

        val drt = dartsThrown.last()
        val roundScore = drt.getGolfScore(currentRound)
        score += roundScore

        currentRound++
    }

    override fun dartThrown(dart: Dart)
    {
        dartsThrown.add(dart)

        val noDarts = dartsThrown.size

        if (noDarts == 3 || dart.getGolfScore(currentRound) <= model.getStopThresholdForDartNo(noDarts))
        {
            finishedRound()
        }
        else
        {
            val dartNo = dartsThrown.size + 1
            val pt = model.throwGolfDart(currentRound, dartNo)
            dartThrown(pt)
        }
    }
}
